bigtable-hbase-1.x-parent/bigtable-hbase-tools/src/main/java/com/google/cloud/bigtable/hbase/tools/HBaseSchemaTranslator.java

```diff
@@ -0,0 +1,414 @@
+    String inputFilePath;
+    String outputFilePath;
+    String tableNameFilter;
+
+    @VisibleForTesting
+    SchemaTranslationOptions() {}
+
+    @VisibleForTesting
+    void validateOptions() {
+      if (outputFilePath != null) {
+        Preconditions.checkArgument(
+            projectId == null && instanceId == null,
+            INSTANCE_ID_KEY + "/" + PROJECT_ID_KEY + " can not be set when output file is set.");
+      } else {
+        Preconditions.checkArgument(
+            projectId != null && instanceId != null, "Schema destination not specified.");
+      }
+
+      if (inputFilePath != null) {
+        Preconditions.checkArgument(
+            zookeeperPort == null && zookeeperQuorum == null,
+            ZOOKEEPER_PORT_KEY
+                + "/"
+                + ZOOKEEPER_QUORUM_KEY
+                + " can not be set when input file is set.");
+        Preconditions.checkArgument(
+            tableNameFilter == null,
+            TABLE_NAME_FILTER_KEY
+                + " is not supported for reading the schema from a table. "
+                + "TableFilter should be used when writing the schema to the file.");
+      } else {
+        Preconditions.checkArgument(
+            zookeeperQuorum != null && zookeeperPort != null, "Schema source not specified. ");
+      }
+    }
+
+    public static SchemaTranslationOptions loadOptionsFromSystemProperties() {
+      SchemaTranslationOptions options = new SchemaTranslationOptions();
+      options.projectId = System.getProperty(PROJECT_ID_KEY);
+      options.instanceId = System.getProperty(INSTANCE_ID_KEY);
+      options.outputFilePath = System.getProperty(OUTPUT_FILE_KEY);
+      options.inputFilePath = System.getProperty(INPUT_FILE_KEY);
+      options.zookeeperQuorum = System.getProperty(ZOOKEEPER_QUORUM_KEY);
+      if (System.getProperty(ZOOKEEPER_PORT_KEY) != null) {
+        options.zookeeperPort = Integer.parseInt(System.getProperty(ZOOKEEPER_PORT_KEY));
+      }
+
+      options.tableNameFilter = System.getProperty(TABLE_NAME_FILTER_KEY);
+
+      // Ensure that the options are set properly
+      // TODO It is possible to validate the options without creating the object, but its less
+      // readable. See if we can make it readable and validate before calling the constructor.
+      try {
+        options.validateOptions();
+      } catch (Exception e) {
+        usage(e.getMessage());
+        throw e;
+      }
+
+      return options;
+    }
+  }
+
+  interface SchemaReader {
+
+    ClusterSchemaDefinition readSchema() throws IOException;
+  }
+
+  static class FileBasedSchemaReader implements SchemaReader {
+
+    String schemaFilePath;
+
+    public FileBasedSchemaReader(String schemaFilePath) {
+      this.schemaFilePath = schemaFilePath;
+    }
+
+    @Override
+    public ClusterSchemaDefinition readSchema() throws IOException {
+      return new ObjectMapper().readValue(new File(schemaFilePath), ClusterSchemaDefinition.class);
+    }
```

Comment:
This is a bit inconsistent ith HBaseSchemaReader. In the HBase one you execute all io in the ctor, here you defer it to readSchema? Is there a reason for this? if so, please document it. If not, please make it consistent

Reply:
I don't do all the IO in HBase reader. I just create the connection and initialize the admin there. The reading still happens in readSchema().